### PR TITLE
Quorum support for MultiMap

### PR DESCRIPTION
**Pull request by tombujok**
EE part: https://github.com/hazelcast/hazelcast-enterprise/pull/1823

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/multimap/impl/MultiMapService.java [hunk 1]
```diff
@@ -88,13 +97,28 @@ public LocalMultiMapStatsImpl createNew(String key) {
...
     private final MultiMapEventsDispatcher dispatcher;
     private final MultiMapEventsPublisher publisher;
+    private final QuorumService quorumService;
+
+    private final ConcurrentMap<String, Object> quorumConfigCache = new ConcurrentHashMap<String, Object>();
+    private final ContextMutexFactory quorumConfigCacheMutexFactory = new ContextMutexFactory();
+    private final ConstructorFunction<String, Object> quorumConfigConstructor = new ConstructorFunction<String, Object>() {
+        @Override
+        public Object createNew(String name) {
+            MultiMapConfig lockConfig = nodeEngine.getConfig().findMultiMapConfig(name);
```
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/multimap/impl/MultiMapService.java]: typo: `lockConfig` -> `multiMapConfig`

File: hazelcast/src/main/java/com/hazelcast/multimap/impl/MultiMapService.java [hunk 2]
```diff
@@ -438,4 +462,16 @@ private Address getReplicaAddress(IPartition partition, int backupCount, int rep
         }
         return replicaAddress;
     }
+
+    @Override
+    public String getQuorumName(String name) {
+        // we use caching here because lock operations are often and we should avoid lock config lookup
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/multimap/impl/MultiMapService.java]: `lock`?

File: hazelcast/src/main/java/com/hazelcast/multimap/impl/operations/ClearBackupOperation.java [hunk 3]
```diff
@@ -19,8 +19,9 @@
 import com.hazelcast.multimap.impl.MultiMapContainer;
 import com.hazelcast.multimap.impl.MultiMapDataSerializerHook;
 import com.hazelcast.spi.BackupOperation;
+import com.hazelcast.spi.impl.MutatingOperation;
 
-public class ClearBackupOperation extends MultiMapOperation implements BackupOperation {
+public class ClearBackupOperation extends MultiMapOperation implements BackupOperation, MutatingOperation {
```
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/multimap/impl/operations/ClearBackupOperation.java]: Is quorum needed for a `BackupOperation`? If primary applies the write, backups should write too. What is the benefit of allowing writes on primary but rejecting a backup?

File: hazelcast/src/main/java/com/hazelcast/multimap/impl/MultiMapService.java (hunk 1 above)
Comment (tombujok) [reply hazelcast/src/main/java/com/hazelcast/multimap/impl/MultiMapService.java]: Fixed.

File: hazelcast/src/main/java/com/hazelcast/multimap/impl/operations/ClearBackupOperation.java (hunk 3 above)
Comment (tombujok) [reply hazelcast/src/main/java/com/hazelcast/multimap/impl/operations/ClearBackupOperation.java]: As discussed on slack.

File: hazelcast/src/main/java/com/hazelcast/multimap/impl/MultiMapService.java (hunk 2 above)
Comment (tombujok) [reply hazelcast/src/main/java/com/hazelcast/multimap/impl/MultiMapService.java]: Fixed.

**Pull request merged**